### ISSUE TITLE
Error: Shape is not valid

### ISSUE THREAD

**Issue by msbeck** [inkstitch/inkstitch#343]
I've attempted tracing several images in Inkscape. Once I've Traced Bitmap, I attempt to Embroider them, but I often get the following:

`path6583: error: shape is not valid.  This can happen if the border crosses over itself.`

It works ok if I have a simple solid object, but not when my converted images have several different colors or separate pieces.

What does it mean by the border? and that it's crossing over itself?

**Comment by wwderw** [comment]
I had this issue on several occasions.  Typically it means that the object isn't closed.  Go to node tool and play around with all the nodes.  More then likely you'll see 2 nodes (if not more) that are right on top of each other and the object isn't closed.

**Comment by lexelby** [comment]
Hi @msbeck, thanks for using Ink/Stitch!  Yes, Trace Bitmap has a tendency to create "invalid" shapes, which can be a pain.  Worse yet, it's usually not easy to see where the shape's border crosses over itself.  It will often have just a little tiny loop that you can't see unless you zoom in really far.

#323 will help with this, once we get it implemented.  In the mean time, if you'd like to post your SVG here (or email to inkstitch at lex dot GD) I'd be happy to take a look and help you make it valid.

BTW, if you're interested in working on #323, I'd be happy to direct you to the relevant bits of the code :)

**Comment by ARKopp** [comment]
I have a issue here with intersecting path. I first had a png and traced it and had problems then I installed ghostscript and saved the eps as svg and still it does not work
[https://www.dropbox.com/s/t4wesrn0xplv7uz/projekt_gefoerdert-eps.svg?dl=0(url)

**Comment by lexelby** [comment]
@ARKopp I think the problem in this case is the same as in #360.  For example, the `j` in `Projekt` is one path with two disconnected shapes in it.

In general though, you're not likely to get good results from simply tracing an image and then running Embroider on it.  As @wwderw hinted at in #349, Ink/Stitch isn't intended as an "autopunch" tool.  "Autopunch" tools try to take any image and convert them into a machine embroidery design automatically, figuring out where to do fill stitch, running stitch, satin stitch, and so on, and what order to do things in.  That's not the use case we're aiming for.

The idea with Ink/Stitch is that you draw out a set of instructions that explain to Ink/Stitch the kinds of stitches you want.  For example, two side-by-side lines with smaller crossed lines between them make a satin stitch.  It looks like a train track or a ladder.  This lets you control the direction that the stitches flow in, while freeing you from having to draw each and every stitch in the satin.

I do sometimes trace bitmaps when creating an Ink/Stitch design, but I really just use that as a starting point.  It's often the case that you need to simplify an image or even re-interpret it in order to make a sensible machine embroidery design that will sew out nicely.

I hope that helps explain things.  Sorry you're having so much trouble!

**Comment by ARKopp** [comment]
What do you normally do with Fonts. I read in the other issue that fill stiches are only recommend with big letters.

**Comment by wwderw** [comment]
Breakdown the shape of the font to be made up of smaller satin columns.

For instances, depending on the size of the "A", you can either have 3 satin columns or 4 satin columns.  The extra column being a cap on the top of the letter.

**Comment by ARKopp** [comment]
i am still new to this satin columns could you give me an example for the A

**Comment by wwderw** [comment]
![satin columns](https://user-images.githubusercontent.com/24878362/49152203-b6966480-f2d7-11e8-9865-44927ab4b545.png)

Ok, the filled in "A" is the original text, so for this example disregard that.

Bottom left is an example of a 3 object version of the "A" and bottom right is an example of a 4 object version of the "A".

Now, while I do have both the same size, each method is going to depend on the size and other production concerns to determine which one to use.

Bottom left is usually used for smaller lettering, we don't want to change up the stitch angle in small areas.  Bottom right is typically used for regular size lettering.

**Comment by lexelby** [comment]
Check out fonts/small_font/*.svg in the inkstitch code for a bunch of examples, one character per layer.

**Comment by Lolasdasdasd** [comment]
I recently started using InkStitch (linux) and it's all working smoothly except for the shape generation which throws - as far as I can tell - arbitrarily the error when I execute 'simulate':
`path4212: error: Fill: Shape is not valid.  This can happen if the border crosses over itself.`
I generate the paths myself using the 'Bezier curves'-tool and Im pretty certain they are closed. I figured out, the error is mostly related to points at the corners of a path and I fix them by moving them a little or making them 'symmetric' nodes. However, when I move the entire object, or minimize its size or copy to a different file, the error reoccurs :( This is really really frustrating. Any clues what's the cause of this?

**Comment by lexelby** [comment]
Hi, @lolasdasdasd!  That error can be annoying.  The latest version of  Ink/Stitch has a new extension, "troubleshoot shapes".  Running that will show you exactly what part of your shape is causing a problem and tell you what the problem is.

It almost always is the border crossing over itself, even if just a tiny amount.  It may be so small you can't see it.  It can help to Simplify the shape or delete extra nodes.

I hope that feels!

**Comment by Lolasdasdasd** [comment]
This is awesome! Thanks so much for the hint!

**Comment by adamsone67** [comment]
Like @Lolasdasdasd I am encountering similar issues. I'm digitizing SVG-based icons, and have become fairly adept at cleaning up a graphic to the point where it will simulate, pretty print and even successfully generate a .PES file.

But... I can drag or cut/paste that exact same source to another location on the page, and all bets are off. I've sometimes fixed this issue by running **Path | Simplify ** but this doesn't always work. I have two hunches so far, the first being that moving a graphic element does not alter its structure, but that it does add an SVG translate tag, and that perhaps this is tripping up the plugin. My other hunch is that some sort of metadata, perhaps CSS metadata, is being automatically added by Inkscape, and that this is causing the problem. No evidence for these yet, just things I'm trying to investigate.

My plan of attack is to create a file containing just some problem content, fixing that content and comparing their respective SVG files. Fingers crossed, the diff will find the culprit.

**Comment by adamsone67** [comment]
Attached are two SVG files -- **works.svg** and **works_no_more.svg**. The working file simulated successfully, the broken version did not.

To create the working file, I "healed" a broken graphic by applying various path operations -- namely **Simplify**, **Break Apart** and **Exclusion** -- to deal with objects having holes and any path-related wonkiness. I saved the working file and saved to the broken filename as well.

To create the broken file, I opened the file saved as the "broken" file in the previous step. First, I confirmed that the content survived the save operation. It still simulated successfully. To break the file, I pressed CTRL+A, CTRL+X, CTRL+ALT+V. The file would no longer simulate.

Diffing the two files, one can see significant discrepancies in data format, including lettercase inconsistency and paired data elements delimited by commas in one case, spaces in the other. The data values themselves don't appear to correlate, in some places. I've attached the files, and the diff itself, for anyone interested. I'm not deep wrt the SVG format spec, so perhaps someone better versed will see something that I wouldn't.
[works_no_more.svg.txt](https://github.com/inkstitch/inkstitch/files/3638233/works_no_more.svg.txt)
[works.svg.txt](https://github.com/inkstitch/inkstitch/files/3638234/works.svg.txt)
[svg-diff.txt](https://github.com/inkstitch/inkstitch/files/3638235/svg-diff.txt)

**Comment by adamsone67** [comment]
Adding that I'm running Ubuntu 19.04, Inkscape 0.92.4, InkStitch 1.26.2

**Comment by adamsone67** [comment]
It's finally worth adding that given the operations breaking my SVG, I doubt that InkStitch is responsible; this almost certainly has to be an issue with Inkscape or a related extension.

**Comment by lexelby** [comment]
Does the "troubleshoot shapes" extension help you find the problem areas of your shapes?

**Comment by adamsone67** [comment]
The following thread may shed some light on the issue I've been encountering. Unfortunately, it's ten years old, so although it offers clues regarding the nature of the error, the workarounds offered therein are no longer applicable as the **Preferences** section has evolved.

http://www.inkscapeforum.com/viewtopic.php?t=5915

**Comment by adamsone67** [comment]
@lexelby It looks fairly slick, and has pointed out problems that became readily apparent upon closer inspection, but this persnickety one it flags as a border crossing issue.. Are the arrowheads intended to point directly at the problem node, or just to indicate the problem path?

**Comment by adamsone67** [comment]
Digging deeper:

Found a better diff tool, am now seeing fairly plainly that during the course of pasting (in place, mind you) the attribute **inkscape:connector-curvature=0** is being injected into some paths and some absolute paths are being converted to relative ones. _(See screenshot #1)_

I'm also seeing that in the broken code, a data value written as **0.001** in the happy file is instead being written using scientific notation -- **10e-4**

![InvalidShape_Analysis-01](https://user-images.githubusercontent.com/45590990/65375095-856c9380-dc5f-11e9-9a4e-b80f4633dcfb.png)
![InvalidShape_Analysis-02](https://user-images.githubusercontent.com/45590990/65375096-856c9380-dc5f-11e9-9f96-aef02aefee79.png)

**Comment by adamsone67** [comment]
After poking around a bit, I found my way over to the **SVG Output** preferences, where the **Path **data** setting was set to **Relative**.

I fixed a broken graphic, changed the setting mentioned above to **Absolute** (see screenshot) and successfully cut/pasted that graphic without breaking it.

@Lolasdasdasd may similarly find that adjusting this setting restores sanity to his workflow.

    Eric

![path-setting](https://user-images.githubusercontent.com/45590990/65375702-80f7a900-dc66-11e9-9642-f6dcbd13ba62.png)

**Comment by lexelby** [comment]
Yes, the arrow head points to the problematic area of the path.  Sometimes there can be multiple nodes stacked so close together that you can't see them, even at maximum magnification.  I deal with this by adding nodes on either side of the problem node(s) and then drag-select and delete the problem node(s).  You can add nodes to a path by double-clicking.

**Comment by lexelby** [comment]
In theory, neither absolute vs relative nor scientific notation should confuse inkstitch.  If you can upload an example broken/fixed SVG, that could help us figure out if this is an ink/stitch bug.

**Comment by adamsone67** [comment]
Attached are the same two files that I posted last night -- this time ZIP'ed, rather than with .TXT extension tacked on. 

[examples.zip](https://github.com/inkstitch/inkstitch/files/3638954/examples.zip)

With **SVG | Path Data** set to **Relative** it should be possible to break a graphic oneself, simply by cutting/pasting it (even pasting in place.) Perhaps quirks peculiar to my SVG content come into play.

Happily, after making that setting change, my frustration level has reduced significantly.

**Comment by tyrosinase** [comment]
Here's another example: Homie works fine here, but if you select all shapes (across layers) and drop him into a different SVG (such as my notebook-cover one) you get a couple of crosses-itself errors which a simple Ctrl++ (single-object "union") will fix, at the cost of having to re-set the Ink/Stitch params 
![IMG_3633](https://user-images.githubusercontent.com/9473609/65391469-87a22100-dd37-11e9-862e-d7e86581d56b.JPG)
[HomerFade4Color.svg.txt](https://github.com/inkstitch/inkstitch/files/3639972/HomerFade4Color.svg.txt)

(Stitchout is on the not-quite-right-green-but-I-have-much-more-of-that vinyl, with Karen-forgot-to-change-out-the-black bobbin thread. D'oh!)

**Comment by lexelby** [comment]
Here's the thing: ink/stitch is doing two things.  First, it converts the bezier curves to a series of small line segments.   It's like taking a circle and approximating it as an octagon.  This is necessary because the shape library can only handle polygons, not bezier curves.

After that, it asks the shape library if the shape is "valid" (whether it crosses itself).  Technically, it's asking about the _approximate_ shape, not the _true_ shape.

It's possible that the approximation process can make a shape that is not "valid" into one that is "valid".  If there's a really tiny loop, the approximation process may just completely omit it.  It's so tiny it won't have any effect on the end result (think smaller than the width of the thread).

It's also possible that the _same shape_, when moved a little, results in a slightly different approximated polygon.  Maybe that one will have its border crossing itself, where the unmoved version didn't.  Computers are always slightly inaccurate when working with decimal points.

Bottom line: Homer probably does have a tiny little loop somewhere on his border.  Normally it gets eliminated in the polygon approximation process, but if you move it or copy-paste it, maybe them it won't be.

**Comment by tyrosinase** [comment]
On 9/22/19 6:24 PM, Lex Neva wrote:
> Bottom line: Homer probably does have a tiny little loop somewhere on
> his border. Normally it gets eliminated in the polygon approximation
> process, but if you move it or copy-paste it, maybe them it won't be.


Not that I've been able to find, at least if the debugger is pointing to
the actual problem vertex. It's pointy (not always the same vertex but
always one of the pointiest of them), but it's definitely not a loop.

**Comment by lexelby** [comment]
Huh.  Well, maybe my theory is wrong... :(

The arrow created by the Troubleshoot extension is indeed supposed to point to the problem area of the shape.  It's possible that there still is a loop there, but it's so tiny you can't see it on the closest zoom level.  It's also possible that there are two or even more vertices there.

 I could add some code to ink/stitch's secret debug mode to have it create an svg that shows the approximate polygon it makes.  That might help us see what's going on here.  I'm not sure when I'll have a chance to do that though.

**Comment by lexelby** [comment]
@tyrosinase what happens if you use the method I described above to isolate and delete that problematic node?

**Comment by tyrosinase** [comment]
On 9/22/19 7:00 PM, Lex Neva wrote:
>
> Huh. Well, maybe my theory is wrong... :(
>
> The arrow created by the Troubleshoot extension is indeed supposed to
> point to the problem area of the shape. It's possible that there still
> is a loop there, but it's so tiny you can't see it on the closest zoom
> level. It's also possible that there are two or even more vertices there.
>
Nope, I had already thought of that.

I also thought about "well, what if it has added an extra 360 degrees
[okay, pi radians] to the angle of something" but that seems improbable
(can't remember offhand how XML stores bezier bits and if that'd even be
possible). I expect it's just getting confused because it's working in
such a confined space there.


> I could add some code to ink/stitch's secret debug mode to have it
> create an svg that shows the approximate polygon it makes. That might
> help us see what's going on here. I'm not sure when I'll have a chance
> to do that though.
>

No rush, it doesn't come up much and it's always work-around-able. Heck,
I bet adding a mid-segment vertex in shapes like that would cut down on
the problem; I think most of those leaf bloops are one continuous segment.

**Comment by tyrosinase** [comment]
On 9/22/19 7:03 PM, Lex Neva wrote:
>
> @tyrosinase <https://github.com/tyrosinase> what happens if you use
> the method I described above to isolate and delete that problematic node?
>

Deleting the sharp point? I do that whenever a satin stitch is involved,
but in this case Homer's small enough I am just running a bunch of
straight stitching so I do want the fill to get all the way out to the
point.


I'll have to see if adding an interim node on the bloops makes a
difference though. I'll do that whenever I get around to playing with
Homer (which will be whenever I find a convenient source of
not-so-blue-green green marine vinyl. In related news, Amazon is now
trying to sell me bilge pump plumbing to go with my imaginary
green-upholstered boat).

**Comment by lexelby** [comment]
I'm suggesting deleting the sharp point only after adding a node on either side of it.  If you do that at a high enough magnification level, it makes no difference to the end result you get from Auto-fill.

**Comment by adamsone67** [comment]
I've yet to fully catch up with this thread, but I just wanted to emphasize that the broken/fixed behavior affecting me was not associated with directly invoking any InkStitch functionality. All I did was cut/paste some code that ran happily in InkStitch before cutting, and not-so-happily after pasting. So, unless InkStitch leverages some hook exposed by Inkscape (I'm not architecturally savvy) I wouldn't expect it to have been involved in that operation.

If my assumptions are correct, I would at worst suspect that InkStitch must adapt to some recent change in Inkscape, so that setting **SVG Path Data** to **Relative** behaves as expected. (If Inkscape is somehow massaging the object out of shape during their cut/paste implementation, the issue might not even lie on InkStitch's end.)

A question I have for those further along:  Does presenting InkStitch with relative vs. absolute path data impose any limitations? I assume that at worst, using one vs. the other will incur extra CPU overhead while processing the file, but does choosing the less efficient approach significantly impede the designer's workflow?

**Comment by lexelby** [comment]
Ink/Stitch should not care one way or another, relative or absolute.  Both methods of specifying an svg path are valid, and Ink/Stitch should be able to parse them both equally.

If you find that that's not the case, I'd love to see an example.

**Comment by pl86** [comment]
I hope my example will help locate the cause.

Invalid Path:
```
m 350.8574,8.2950714 -0.67138,28.4667996 -20.00733,0.805665 -5.10254,63.647464 -28.33252,1.47705 0.13427,-63.647463 -20.00732,1.074218 0.80566,-31.6894562 73.18116,-0.1342774 z
```

I only changed last point `73.18116` to `73.18115` and it became Valid!
```
m 350.8574,8.2950714 -0.67138,28.4667996 -20.00733,0.805665 -5.10254,63.647464 -28.33252,1.47705 0.13427,-63.647463 -20.00732,1.074218 0.80566,-31.6894562 73.18115,-0.1342774 z
```

There may be something wrong with the calculations.

**Comment by lexelby** [comment]
@pl86 Sorry it took me so long to reply!  I tested your path, and even the first one (with `73.18116`) showed up as valid for me.  I'm not sure how else to try to reproduce your problem, other than maybe having you post your entire SVG?

**Comment by stevensteichen** [comment]
I have a technical/developer background and went in hunter for the root cause as well. 

I have a path that looks like the letter R. It reports a Border Crossing Itself Error. From now on, shorted to Border Error, BE.

This R path, can't be fixed, I've broken, merged, moved, replace...The BE just moves up a node and chasing it around the entire path the BE returns to the original node.

I have all the xml to reproduce if anyone wants the xml string, I'll attach them.

I'm working with the XML as text so that I control, exactly what changes, and I can fix all(8) of the BE in my file by removing the style. Small scale tests on the R path I can add and remove all the BE with a simple style change. The node XML is unchanged.

If I set Fill to 'none. The path(s) is correct, no BE. Setting Stroke to any Width is still correct. However, any Fill throws a BE, this is with or without any Stroke set. This in not a node issue that can be corrected that I've ever read as fill can't cross itself as fill is not a path. I get that if there exist a tiny loop it would be an issue but that would be on the path and all the nodes on the path are valid if Troubleshoot Object is to be trusted. It's always reported as a node issue with the path. However, the path is valid. it's the fill (style attribute) that is interpreted as crossing itself in my case.

If anyone can explain why Fill is a Border, I'd be very interested.

**Comment by pl86** [comment]
Can't upload svg, so I changed extension to txt:

[border-crossing-itself--oval--v1--invalid.txt](https://github.com/inkstitch/inkstitch/files/5689443/border-crossing-itself--oval--v1--invalid.txt)

This is just oval with fill and stroke. It reports a Border Crossing Itself Error. I've tested 3 versions of Inkscape (0.91, 1.0.0beta2 (2b71d25) and 1.0.1 (c497b03c)) and 2 versions of ink/stitch extension (1.26.2 and 1.29.1)

Path attributes:
```
style="fill:#000000;stroke:#000000;stroke-width:0.264583"
d="M 101.91654,19.844437 A 21.356897,2.8514659 0 0 1 80.559639,22.695903 21.356897,2.8514659 0 0 1 59.202739,19.844437 21.356897,2.8514659 0 0 1 80.559639,16.992971 21.356897,2.8514659 0 0 1 101.91654,19.844437 Z" />
```

Tool: Effects -> Ink/Stitch -> Fill Tools -> Break Apart Fill Objects... fixed the bug, but it covert round shape to lines like: Modify Path -> Flatten Beziers

![image](https://user-images.githubusercontent.com/1517916/102090526-de54b880-3e1d-11eb-9584-c4df2ba9c5e7.png)

It also adds transform attribute to path element:
```
d="M385.196416593,75.0026053012  385.097213721,74.4681949876  ...  385.196416593,75.0026053012  Z"
transform="matrix(0.264583,-0.000000,-0.000000,0.264583,-0.000000,-0.000000)" />
```

The numbers have changed so the error is related to the calculations or maybe it is in some svg path libraries.

I did some additional tests.

1. remove stroke - still **INVALID**

2. remove fill - **VALID**

3. I tried a similar method as before - a small change in numbers. Just remove last digit from path - **VALID**
`...101.91654,19.844437 Z" />` to `...101.91654,19.84443 Z" />`

4. Modify Path -> Flatten Beziers - still **INVALID**

5. I clicked on the point indicated by the red arrow from "Border crosses itself" and just hit Left Arrow and then Right to make it back to its original position. **VALID**.
Path attribute changed to:
```
M 101.91654,19.844437 C 101.91654,21.419258 92.354729,22.695903 80.559639,22.695903 68.764549,22.695903 59.202739,21.419258 59.202739,19.844437 59.202739,18.269616 68.764549,16.992971 80.559639,16.992971 92.354729,16.992971 101.91654,18.269616 101.91654,19.844437 Z
```
